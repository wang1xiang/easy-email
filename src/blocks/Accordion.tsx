import { omit } from 'lodash';
import { BasicType } from '@/constants';
import { RecursivePartial } from '@/typings';
import React from 'react';
import { IAccordion } from '@/components/core/blocks/basic/Accordion';
import MjmlBlock from '@/components/core/MjmlBlock';

export type AccordionProps = RecursivePartial<IAccordion['data']> &
  RecursivePartial<IAccordion['attributes']> & {
    children?: JSX.Element | JSX.Element[] | string;
  };

export function Accordion(props: AccordionProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children'])}
      value={props.value}
      type={BasicType.ACCORDION}
    >
      {props.children}
    </MjmlBlock>
  );
}
