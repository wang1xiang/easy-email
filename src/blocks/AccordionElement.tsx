import { omit } from 'lodash';
import { BasicType } from '@/constants';
import { RecursivePartial } from '@/typings';
import React from 'react';
import { IAccordionElement } from '@/components/core/blocks/basic/AccordionElement';
import MjmlBlock from '@/components/core/MjmlBlock';

export type AccordionElementProps = RecursivePartial<
  IAccordionElement['data']
> &
  RecursivePartial<IAccordionElement['attributes']> & {
    children?: JSX.Element | JSX.Element[] | string;
  };

export function AccordionElement(props: AccordionElementProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children'])}
      value={props.value}
      type={BasicType.ACCORDION_ELEMENT}
    >
      {props.children}
    </MjmlBlock>
  );
}
