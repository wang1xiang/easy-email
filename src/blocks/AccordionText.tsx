import { omit } from 'lodash';
import { BasicType } from '@/constants';
import { RecursivePartial } from '@/typings';
import React from 'react';
import { IAccordionText } from '@/components/core/blocks/basic/AccordionText';
import MjmlBlock from '@/components/core/MjmlBlock';

export type AccordionTextProps = RecursivePartial<IAccordionText['data']> &
  RecursivePartial<IAccordionText['attributes']> & {
    children?: JSX.Element | JSX.Element[] | string;
  };

export function AccordionText(props: AccordionTextProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children'])}
      value={props.value}
      type={BasicType.ACCORDION_TEXT}
    >
      {props.children}
    </MjmlBlock>
  );
}
