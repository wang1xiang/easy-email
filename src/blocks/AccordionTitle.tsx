import { omit } from 'lodash';
import { BasicType } from '@/constants';
import { RecursivePartial } from '@/typings';
import React from 'react';
import { IAccordionTitle } from '@/components/core/blocks/basic/AccordionTitle';
import MjmlBlock from '@/components/core/MjmlBlock';

export type AccordionTitleProps = RecursivePartial<IAccordionTitle['data']> &
  RecursivePartial<IAccordionTitle['attributes']> & {
    children?: JSX.Element | JSX.Element[] | string;
  };

export function AccordionTitle(props: AccordionTitleProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children'])}
      value={props.value}
      type={BasicType.ACCORDION_TITLE}
    >
      {props.children}
    </MjmlBlock>
  );
}
