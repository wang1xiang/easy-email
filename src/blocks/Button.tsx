import { omit } from 'lodash';
import { BasicType } from '@/constants';
import { RecursivePartial } from '@/typings';
import React from 'react';
import { IButton } from '@/components/core/blocks/basic/Button';
import MjmlBlock from '@/components/core/MjmlBlock';

export type ButtonProps = RecursivePartial<IButton['data']> &
  RecursivePartial<IButton['attributes']> & {
    children?: JSX.Element | JSX.Element[] | string;
  };

export function Button(props: ButtonProps) {
  return (
    <MjmlBlock
      attributes={omit(props, ['data', 'children'])}
      value={props.value}
      type={BasicType.BUTTON}
    >
      {props.children}
    </MjmlBlock>
  );
}
