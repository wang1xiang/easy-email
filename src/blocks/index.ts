export { Page } from './Page'
export { Section } from './Section'
export { Column } from './Column'
export { Text } from './Text'
export { Image } from './Image'
export { Group } from './Group'
export { Button } from './Button'
export { Divider } from './Divider'
export { Wrapper } from './Wrapper'
export { Spacer } from './Spacer'
export { Raw } from './Raw'
export { Accordion } from './Accordion'
export { AccordionElement } from './AccordionElement'
export { AccordionTitle } from './AccordionTitle'
export { AccordionText } from './AccordionText'
export { Carousel } from './Carousel'
export { Hero } from './Hero'
export { Navbar } from './Navbar'
export { Social } from './Social'
export { Table } from './Table'
