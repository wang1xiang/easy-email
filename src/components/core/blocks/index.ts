import { Page } from './basic/Page';
import { Section } from './basic/Section';
import { Column } from './basic/Column';
import { Text } from './basic/Text';
import { Image } from './basic/Image';
import { Group } from './basic/Group';
import { Button } from './basic/Button';
import { Divider } from './basic/Divider';
import { Wrapper } from './basic/Wrapper';
import { Spacer } from './basic/Spacer';

import { Raw } from './basic/Raw';
import { Accordion } from './basic/Accordion';
import { AccordionElement } from './basic/AccordionElement';
import { AccordionTitle } from './basic/AccordionTitle';
import { AccordionText } from './basic/AccordionText';
import { Carousel } from './basic/Carousel';
import { Table } from './basic/Table';
import { Hero } from './basic/Hero';
import { Navbar } from './basic/Navbar';
import { Social } from './basic/Social';
import { IBlock } from '@/typings';
import { BasicType, BlockType } from '@/constants';

const basicBlocks = {
  Page,
  Section,
  Column,
  Text,
  Image,
  Group,
  Button,
  Divider,
  Wrapper,
  Spacer,
  Raw,
  Accordion,
  AccordionElement,
  AccordionTitle,
  AccordionText,
  Carousel,
  Hero,
  Navbar,
  Social,
  // TODO:
  Table,
};

export class BlocksMap {
  static basicBlocksMap = basicBlocks;
  static externalBlocksMap: { [key: string]: IBlock; } = {};
  private static autoCompletePath: { [key: string]: Array<BlockType[]>; } = {};

  static getBlocks() {
    return [
      ...Object.values(this.basicBlocksMap),
      ...Object.values(this.externalBlocksMap),
    ];
  }

  static registerBlocks(blocksMap: { [key: string]: IBlock; }) {
    Object.assign(this.externalBlocksMap, blocksMap);
    this.autoCompletePath = this.setAutoCompletePath();
  }

  static findBlockByType(type: string): IBlock {
    return this.getBlocks().find((child) => {
      return child?.type === type;
    }) as any;
  }

  static findBlocksByType(types: Array<BlockType>): IBlock[] {
    return types.map((item) => {
      const block = this.getBlocks().find((child) => {
        return child.type === item;
      });
      if (!block) {
        throw new Error(`Cannot find ${item}`);
      }
      return block;
    });
  }

  static getBlock<
    E extends { [key: string]: IBlock; },
    B extends typeof BlocksMap.basicBlocksMap,
    A extends B & E,
    T extends keyof A
  >(name: T): A[T] {
    const key: any = name;
    return this.basicBlocksMap[key] || this.externalBlocksMap[key];
  }

  static setAutoCompletePath() {
    const paths: { [key: string]: Array<BlockType[]>; } = {};

    const renderFullPath = (
      type: BlockType,
      pathObj: Array<BlockType[]>,
      prevPaths: BlockType[]
    ) => {
      const block = this.findBlockByType(type);
      const currentPaths = [...prevPaths, type];
      if (block.validParentType.length === 0) {
        pathObj.push(currentPaths);
      }
      return block.validParentType.map((item) => {
        return renderFullPath(item, pathObj, currentPaths);
      });
    };

    this.getBlocks().forEach((item) => {
      paths[item.type] = [];
      renderFullPath(item.type, paths[item.type], []);
    });
    return paths;
  }

  static getAutoCompleteFullPath() {
    if (Object.keys(this.autoCompletePath).length === 0) {
      this.autoCompletePath = this.setAutoCompletePath();
    }
    return this.autoCompletePath;
  }

  static getAutoCompletePath(type: BlockType, targetType: BlockType) {
    const block = BlocksMap.findBlockByType(type);
    if (block.validParentType.includes(targetType)) {
      return [];
    }
    const paths = this.getAutoCompleteFullPath()[type].find((item) =>
      item.filter((_, index) => index !== 0).includes(targetType)
    );

    if (!paths) return null;
    const findIndex = paths.findIndex((item) => item === targetType);
    return paths.slice(1, findIndex);
  }
}
