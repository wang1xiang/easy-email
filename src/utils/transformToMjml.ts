import { parseMjmlBlockToBlockData } from '@/utils/parseMjmlBlockToBlockData';
import { isValidElement } from 'react';
import { BlocksMap } from '@/components/core/blocks';
import { ICarousel } from '@/components/core/blocks/basic/Carousel';
import { INavbar } from '@/components/core/blocks/basic/Navbar';
import { IPage } from '@/components/core/blocks/basic/Page';
import { ISocial } from '@/components/core/blocks/basic/Social';
import { BasicType, BlockType } from '@/constants';
import { IBlockData } from '@/typings';
import { pickBy, identity, isObject, isBoolean, isString } from 'lodash';
import {
  getChildIdx,
  getNodeIdxClassName,
  getNodeTypeClassName,
} from './block';
import { classnames } from './classnames';

export type TransformToMjmlOption =
  | {
    data: IBlockData;
    idx: string | null; // current idx
    context: IBlockData;
    mode: 'testing';
  }
  | {
    idx?: string | null; // current idx, default page idx
    data: IBlockData;
    context: IBlockData;
    mode: 'production';
  };

export function transformToMjml(options: TransformToMjmlOption): string {
  const {
    data,
    idx = 'content',
    context = data,
    mode = 'production',
  } = options;
  if ((isBoolean(data?.data?.hidden) && data?.data?.hidden) || data?.data?.hidden === 'true') {
    return '';
  }

  const att = pickBy(
    {
      ...data.attributes,
    },
    identity
  );

  const isTest = mode === 'testing';
  const placeholder = isTest && idx ? renderPlaceholder(data.type) : '';

  if (isTest && idx) {
    att['css-class'] = classnames(
      att['css-class'],
      'email-block',
      getNodeIdxClassName(idx),
      getNodeTypeClassName(data.type)
    );
  }

  if (data.type === BasicType.PAGE) {
    att['css-class'] = classnames(att['css-class'], 'mjml-body');
  }

  const attributeStr = Object.keys(att)
    .filter((key) => att[key] !== '') // filter att=""
    .map((key) => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const val = isString(att[key]) ? att[key].replace(/"/gm, '') : att[key];
      return `${key}="${val}"`;
    })
    .join(' ');

  const block = BlocksMap.findBlockByType(data.type);

  if (!block) {
    throw new Error(
      `Can not find ${data.type} block!!! Have you registered this block ?`
    );
  }

  if (block.render) {
    const transformBlockData = block.render(data, idx, context);
    const transformData = isValidElement(transformBlockData)
      ? parseMjmlBlockToBlockData(transformBlockData)
      : transformBlockData;
    att['css-class'] = classnames(att['css-class'], transformData['css-class']);
    return transformToMjml({
      data: {
        ...transformData,
        attributes: {
          ...transformData.attributes,
          'css-class': att['css-class'],
        },
      },
      idx: data.children.length > 0 ? idx : null,
      context,
      mode,
    });
  }

  const children = data.children
    .map((child, index) =>
      transformToMjml({
        data: child,
        idx: idx ? getChildIdx(idx, index) : null,
        context,
        mode,
      })
    )
    .join('\n');

  switch (data.type) {
    case BasicType.PAGE:
      const metaData = generaMjmlMetaData(data);
      const value: IPage['data']['value'] = data.data.value;

      const breakpoint = value.breakpoint
        ? `<mj-breakpoint width="${data.data.value.breakpoint}" />`
        : '';

      const nonResponsive = !value.responsive
        ? `<mj-raw>
            <meta name="viewport" />
           </mj-raw>
           <mj-style inline="inline">.mjml-body { width: ${data.attributes.width || '600px'
        }; margin: 0px auto; }</mj-style>`
        : '';
      const styles =
        value.headStyles
          ?.map(
            (style) =>
              `<mj-style ${style.inline ? 'inline="inline"' : ''}>${style.content
              }</mj-style>`
          )
          .join('\n') || '';

      return `
        <mjml>
          <mj-head>
              ${metaData}
              ${nonResponsive}
              ${styles}
              ${breakpoint}
            <mj-attributes>
              ${value.headAttributes}
              ${value['font-family']
          ? `<mj-all font-family="${value['font-family']}" />`
          : ''
        }
              ${value['font-size']
          ? `<mj-text font-size="${value['font-size']}" />`
          : ''
        }
              ${value['text-color']
          ? `<mj-text color="${value['text-color']}" />`
          : ''
        }
              ${value['line-height']
          ? `<mj-text line-height="${value['line-height']}" />`
          : ''
        }
              ${value['content-background-color']
          ? `<mj-wrapper background-color="${value['content-background-color']}" />
                     <mj-section background-color="${value['content-background-color']}" />
                    `
          : ''
        }
              ${value.fonts
          ?.filter(Boolean)
          .map(
            (item) =>
              `<mj-font name="${item.name}" href="${item.href}" />`
          )}
            </mj-attributes>
          </mj-head>
          <mj-body ${attributeStr}>
            ${children}
          </mj-body>
        </mjml>
        `;
    case BasicType.COLUMN:
      return `
              <mj-column ${attributeStr}>
               ${children || placeholder}
              </mj-column>
            `;
    case BasicType.SECTION:
      return `
              <mj-section ${attributeStr}>
               ${children || `<mj-column>${placeholder}</mj-column>`}
              </mj-section>
            `;
    case BasicType.GROUP:
      return `
              <mj-group ${attributeStr}>
               ${children || `<mj-column>${placeholder}</mj-column>`}
              </mj-group>
            `;
    case BasicType.WRAPPER:
      return `
              <mj-wrapper ${attributeStr}>
               ${children ||
        `<mj-section><mj-column>${placeholder}</mj-column></mj-section>`
        }
              </mj-wrapper>
            `;
    case BasicType.CAROUSEL:
      const carouselImages = (data as ICarousel).data.value.images
        .map((image, index) => {
          const imageAttributeStr = Object.keys(image)
            .filter((key) => key !== 'content' && att[key] !== '') // filter att=""
            .map((key) => `${key}="${image[key]}"`)
            .join(' ');
          return `
          <mj-carousel-image ${imageAttributeStr} />
          `;
        })
        .join('\n');

      return `
        <mj-carousel ${attributeStr}>
         ${carouselImages}
        </mj-carousel>
      `;
    case BasicType.NAVBAR:
      const links = (data as INavbar).data.value.links
        .map((link, index) => {
          const linkAttributeStr = Object.keys(link)
            .filter((key) => key !== 'content' && att[key] !== '') // filter att=""
            .map((key) => `${key}="${link[key]}"`)
            .join(' ');
          return `
          <mj-navbar-link ${linkAttributeStr}>${link.content}</mj-navbar-link>
          `;
        })
        .join('\n');
      return `
              <mj-navbar ${attributeStr}>
               ${links}
              </mj-navbar>
            `;
    case BasicType.SOCIAL:
      const elements = (data as ISocial).data.value.elements
        .map((element, index) => {
          const elementAttributeStr = Object.keys(element)
            .filter((key) => key !== 'content' && att[key] !== '') // filter att=""
            .map((key) => `${key}="${element[key]}"`)
            .join(' ');
          return `
          <mj-social-element ${elementAttributeStr}>${element.content}</mj-social-element>
          `;
        })
        .join('\n');
      return `
              <mj-social ${attributeStr}>
               ${elements}
              </mj-social>
            `;
    case BasicType.RAW:
      return `
              <mj-raw ${attributeStr}>
                ${data.data.value?.content}
              </mj-raw>
            `;
    default:
      return `
          <mj-${data.type} ${attributeStr}>
           ${children || data.data.value?.content || ''}
          </mj-${data.type}>
        `;
  }
}

export function renderPlaceholder(type: BlockType) {
  let text = '';
  if (type === BasicType.PAGE) {
    text = 'Drop a Wrapper block here';
  } else if (type === BasicType.WRAPPER) {
    text = 'Drop a Section block here';
  } else if (type === BasicType.SECTION || type === BasicType.GROUP) {
    text = 'Drop a Column block here';
  } else if (type === BasicType.COLUMN) {
    text = 'Drop a content block here';
  }

  return `
   <mj-text color="#666">
    <div style="text-align: center">
      <div>
        <svg width="300" fill="currentColor" style="max-width: 100%;" viewBox="-20 -5 80 60">
          <g>
            <path d="M23.713 23.475h5.907c.21 0 .38.17.38.38v.073c0 .21-.17.38-.38.38h-5.907a.38.38 0 0 1-.38-.38v-.073c0-.21.17-.38.38-.38zm.037-2.917h9.167a.417.417 0 0 1 0 .834H23.75a.417.417 0 0 1 0-.834zm0-2.5h9.167a.417.417 0 0 1 0 .834H23.75a.417.417 0 0 1 0-.834zm-.037-3.333h5.907c.21 0 .38.17.38.38v.073c0 .21-.17.38-.38.38h-5.907a.38.38 0 0 1-.38-.38v-.073c0-.21.17-.38.38-.38zm.037-2.917h9.167a.417.417 0 0 1 0 .834H23.75a.417.417 0 0 1 0-.834zm0-2.916h9.167a.417.417 0 0 1 0 .833H23.75a.417.417 0 0 1 0-.833zm-3.592 8.75a.675.675 0 0 1 .675.691v6.142c0 .374-.3.679-.675.683h-6.15a.683.683 0 0 1-.675-.683v-6.142a.675.675 0 0 1 .675-.691h6.15zM20 24.308v-5.833h-5.833v5.833H20zm.158-15.833a.675.675 0 0 1 .675.692v6.141c0 .374-.3.68-.675.684h-6.15a.683.683 0 0 1-.675-.684V9.167a.675.675 0 0 1 .675-.692h6.15zM20 15.142V9.308h-5.833v5.834H20zM37.167 0A2.809 2.809 0 0 1 40 2.833V30.5a2.809 2.809 0 0 1-2.833 2.833h-3.834v3H32.5v-3h-23A2.808 2.808 0 0 1 6.667 30.5v-23H3.583v-.833h3.084V2.833A2.808 2.808 0 0 1 9.5 0h27.667zm2 30.5V2.833a2.025 2.025 0 0 0-2-2H9.5a2.025 2.025 0 0 0-2 2V30.5a2.025 2.025 0 0 0 2 2h27.667a2.025 2.025 0 0 0 2-2zM0 27.75h.833V31H0v-3.25zm0-13h.833V18H0v-3.25zm0 22.833V34.25h.833v3.25L0 37.583zM0 21.25h.833v3.25H0v-3.25zM2.583 40l.084-.833h3.166V40h-3.25zm27.917-.833c.376.006.748-.08 1.083-.25l.417.666a2.875 2.875 0 0 1-1.5.417h-1.833v-.833H30.5zm-8.333 0h3.25V40h-3.25v-.833zm-6.584 0h3.25V40h-3.25v-.833zm-6.5 0h3.25V40h-3.25v-.833zM0 9.5c.01-.5.154-.99.417-1.417l.666.417c-.17.305-.256.65-.25 1v2H0v-2z"></path>
          </g>
          <text x="-16" y="50" font-size="5px">${text}</text>
        </svg>
      </div>
    </div>
   </mj-text>
  `;
}

export function generaMjmlMetaData(data: IPage) {
  const values = data.data.value;
  const attributes = [
    'content-background-color',
    'text-color',
    'font-family',
    'font-size',
    'line-height',
    'user-style',
    'responsive',
  ];
  return `
    <mj-html-attributes>
      ${attributes
      .filter((key) => values[key] !== undefined)
      .map((key) => {
        const isMultipleAttributes = isObject(values[key]);
        const value = isMultipleAttributes
          ? Object.keys(values[key])
            .map((childKey) => `${childKey}="${values[key][childKey]}"`)
            .join(' ')
          : `${key}="${values[key]}"`;
        return `<mj-html-attribute class="easy-email" multiple-attributes="${isMultipleAttributes}" attribute-name="${key}" ${value}></mj-html-attribute>`;
      })
      .join('\n')}

    </mj-html-attributes>
  `;
}
