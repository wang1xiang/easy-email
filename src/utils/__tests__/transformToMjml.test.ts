import { BlocksMap } from '@/components/core/blocks';
import { transformToMjml } from '../transformToMjml';

describe('Test transformToMjml when responsive is "false"', () => {
  const content = BlocksMap.basicBlocksMap.Page.create({
    data: {
      value: {
        responsive: false,
      },
    },
    children: [
      BlocksMap.basicBlocksMap.Section.create({
        children: [
          BlocksMap.basicBlocksMap.Column.create({
            children: [BlocksMap.basicBlocksMap.Text.create({})],
          }),
        ],
      }),
    ],
  });

  const parseHtml = transformToMjml({
    data: content,
    mode: 'production',
    context: content,
  });

  it('should contains the mark of "<meta name="viewport" />"', () => {
    expect(parseHtml).toContain('<meta name="viewport" />');
  });
});

describe('Test transformToMjml when responsive is "true"', () => {
  const content = BlocksMap.basicBlocksMap.Page.create({
    data: {
      value: {
        responsive: true,
      },
    },
    children: [
      BlocksMap.basicBlocksMap.Section.create({
        children: [
          BlocksMap.basicBlocksMap.Column.create({
            children: [BlocksMap.basicBlocksMap.Text.create({})],
          }),
        ],
      }),
    ],
  });

  const parseHtml = transformToMjml({
    data: content,
    mode: 'production',
    context: content,
  });

  it('should contains the mark of responsive="true"', () => {
    expect(
      parseHtml.includes('attribute-name="responsive" responsive="true"')
    ).toBeTruthy();
  });
});

describe('Test transformToMjml when mode is "testing"', () => {
  const content = BlocksMap.basicBlocksMap.Page.create({
    children: [
      BlocksMap.basicBlocksMap.Section.create({
        children: [
          BlocksMap.basicBlocksMap.Column.create({
            children: [BlocksMap.basicBlocksMap.Text.create({})],
          }),
        ],
      }),
    ],
  });

  const parseHtml = transformToMjml({
    data: content,
    mode: 'testing',
    context: content,
    idx: 'content',
  });

  it('should contains node-type-page', () => {
    expect(parseHtml).toContain('node-type-page');
  });

  it('should render as expected', () => {
    expect(parseHtml).toMatchSnapshot();
  });
});

describe('Test transformToMjml when mode is "production"', () => {
  const content = BlocksMap.basicBlocksMap.Page.create({
    children: [
      BlocksMap.basicBlocksMap.Section.create({
        children: [
          BlocksMap.basicBlocksMap.Column.create({
            children: [BlocksMap.basicBlocksMap.Text.create({})],
          }),
        ],
      }),
    ],
  });

  const parseHtml = transformToMjml({
    data: content,
    mode: 'production',
    context: content,
  });

  it('should not contains node-type-page', () => {
    expect(parseHtml.includes('node-type-page')).toBeFalsy();
  });

  it('should render as expected', () => {
    expect(parseHtml).toMatchSnapshot();
  });
});
