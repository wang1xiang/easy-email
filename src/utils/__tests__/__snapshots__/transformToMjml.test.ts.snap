// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test transformToMjml when mode is "production" should render as expected 1`] = `
"
        <mjml>
          <mj-head>
              
    <mj-html-attributes>
      <mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"text-color\\" text-color=\\"#000000\\"></mj-html-attribute>
<mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"font-family\\" font-family=\\"lucida Grande,Verdana,Microsoft YaHei\\"></mj-html-attribute>
<mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"font-size\\" font-size=\\"14px\\"></mj-html-attribute>
<mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"line-height\\" line-height=\\"1.7\\"></mj-html-attribute>
<mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"responsive\\" responsive=\\"true\\"></mj-html-attribute>

    </mj-html-attributes>
  
              
              
              <mj-breakpoint width=\\"480px\\" />
            <mj-attributes>
              
              <mj-all font-family=\\"lucida Grande,Verdana,Microsoft YaHei\\" />
              <mj-text font-size=\\"14px\\" />
              <mj-text color=\\"#000000\\" />
              <mj-text line-height=\\"1.7\\" />
              
              
            </mj-attributes>
          </mj-head>
          <mj-body background-color=\\"#efeeea\\" width=\\"600px\\" css-class=\\"mjml-body\\">
            
              <mj-section padding=\\"20px 0px 20px 0px\\" border=\\"none\\" direction=\\"ltr\\" text-align=\\"center\\" background-repeat=\\"repeat\\" background-size=\\"auto\\" background-position=\\"top center\\">
               
              <mj-column padding=\\"0px 0px 0px 0px\\" border=\\"none\\" vertical-align=\\"top\\">
               
          <mj-text padding=\\"10px 25px 10px 25px\\" align=\\"left\\">
           Make it easy for everyone to compose emails!
          </mj-text>
        
              </mj-column>
            
              </mj-section>
            
          </mj-body>
        </mjml>
        "
`;

exports[`Test transformToMjml when mode is "testing" should render as expected 1`] = `
"
        <mjml>
          <mj-head>
              
    <mj-html-attributes>
      <mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"text-color\\" text-color=\\"#000000\\"></mj-html-attribute>
<mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"font-family\\" font-family=\\"lucida Grande,Verdana,Microsoft YaHei\\"></mj-html-attribute>
<mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"font-size\\" font-size=\\"14px\\"></mj-html-attribute>
<mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"line-height\\" line-height=\\"1.7\\"></mj-html-attribute>
<mj-html-attribute class=\\"easy-email\\" multiple-attributes=\\"false\\" attribute-name=\\"responsive\\" responsive=\\"true\\"></mj-html-attribute>

    </mj-html-attributes>
  
              
              
              <mj-breakpoint width=\\"480px\\" />
            <mj-attributes>
              
              <mj-all font-family=\\"lucida Grande,Verdana,Microsoft YaHei\\" />
              <mj-text font-size=\\"14px\\" />
              <mj-text color=\\"#000000\\" />
              <mj-text line-height=\\"1.7\\" />
              
              
            </mj-attributes>
          </mj-head>
          <mj-body background-color=\\"#efeeea\\" width=\\"600px\\" css-class=\\"email-block node-idx-content node-type-page mjml-body\\">
            
              <mj-section padding=\\"20px 0px 20px 0px\\" border=\\"none\\" direction=\\"ltr\\" text-align=\\"center\\" background-repeat=\\"repeat\\" background-size=\\"auto\\" background-position=\\"top center\\" css-class=\\"email-block node-idx-content.children.[0] node-type-section\\">
               
              <mj-column padding=\\"0px 0px 0px 0px\\" border=\\"none\\" vertical-align=\\"top\\" css-class=\\"email-block node-idx-content.children.[0].children.[0] node-type-column\\">
               
          <mj-text padding=\\"10px 25px 10px 25px\\" align=\\"left\\" css-class=\\"email-block node-idx-content.children.[0].children.[0].children.[0] node-type-text\\">
           Make it easy for everyone to compose emails!
          </mj-text>
        
              </mj-column>
            
              </mj-section>
            
          </mj-body>
        </mjml>
        "
`;
